types/rrdtool/index.d.ts

```diff
@@ -0,0 +1,142 @@
+    | ''
+    | `:${number}`}`;
+
+// RRA:DEVPREDICT:rows:rra-num
+type RoundRobinArchiveDevPredict = `RRA:DEVPREDICT:${Duration}:${number}`;
+
+// RRA:FAILURES:rows:threshold:window length:rra-num
+type RoundRobinArchiveFailures = `RRA:FAILURES:${Duration}:${number}:${number}:${number}`;
+
+/**
+ * https://oss.oetiker.ch/rrdtool/doc/rrdcreate.en.html
+ *
+ * Round robin archives:
+ *  - `"RRA:{AVERAGE | MIN | MAX | LAST}:xff:steps:rows"`
+ *  - `"RRA:HWPREDICT:rows:alpha:beta:seasonal period[:rra-num]"`
+ *  - `"RRA:MHWPREDICT:rows:alpha:beta:seasonal period[:rra-num]"`
+ *  - `"RRA:SEASONAL:seasonal period:gamma:rra-num[:smoothing-window]"`
+ *  - `"RRA:DEVSEASONAL:seasonal period:gamma:rra-num[:smoothing-window]"`
+ *  - `"RRA:DEVPREDICT:rows:rra-num"`
+ *  - `"RRA:FAILURES:rows:threshold:window length:rra-num"`
+ *
+ * Variable types:
+ *  - `xff`: number (between 0 and 1)
+ *  - `steps`: number (seconds) OR duration string
+ *  - `rows`: number (seconds) OR duration string
+ *  - `alpha`: number (between 0 and 1)
+ *  - `beta`: number (between 0 and 1)
+ *  - `gamma`: number (between 0 and 1)
+ *  - `seasonal period`: number (seconds) OR duration string
+ *  - `rra-num`: number (1-based RRA index)
+ *  - `smoothing-window`: number (between 0 and 1)
+ */
+export type RoundRobinArchive =
+    | RoundRobinArchive_
+    | RoundRobinArchiveHWPredict
+    | RoundRobinArchiveSeasonal
+    | RoundRobinArchiveDevPredict
+    | RoundRobinArchiveFailures;
+
+// XXX: Not really correct, this would be better but does not work correctly (?):
+// (error: Error, data: undefined) => void | ((error: null, data: Array<DataPoint<A>>) => void)
+type FetchCallback<D extends RrdtoolData> = (error: Error, data: Array<RrdtoolDataPoint<D>>) => void;
+
+export interface RrdtoolDatabase<D extends RrdtoolData = any> {
+    /**
+     * Insert data into the database.
+     */
+    update(ts: Timestamp, values: Partial<D>, cb?: (error: Error) => void): void;
+    update(values: D, cb?: (error: Error) => void): void;
+
+    /**
+     * Fetch a span of data from the database.
+     */
+    fetch(cf: ConsolidationFunction, start: Timestamp, stop: Timestamp, res: number, cb: FetchCallback<D>): void;
+    fetch(cf: ConsolidationFunction, start: Timestamp, stop: Timestamp, cb: FetchCallback<D>): void;
+}
+
+export type RrdtoolArgument<D extends RrdtoolData> = DataSource<D> | RoundRobinArchive;
+
+declare namespace rrdtool {
+    /**
+     * Creates a new database.
+     */
+    function create<D extends RrdtoolData>(
+        file: string,
+        opts: {
+            step?: Duration;
+            start?: Timestamp;
+            force?: boolean;
+        },
+        args: ReadonlyArray<RrdtoolArgument<D>>,
+    ): RrdtoolDatabase<D>;
+
+    /**
+     * Loads an existing database.
+     */
+    function open(file: string): RrdtoolDatabase;
+
+    /**
+     * Returns the current unix timestamp.
+     */
+    function now(): number;
+}
+
+export default rrdtool;
```

Comment:
You could probably move these functions out of a namespace and `export` them separately instead:
```suggestion
/**
 * Creates a new database.
 */
export function create<D extends RrdtoolData>(
    file: string,
    opts: {
        step?: Duration;
        start?: Timestamp;
        force?: boolean;
    },
    args: ReadonlyArray<RrdtoolArgument<D>>,
): RrdtoolDatabase<D>;

/**
 * Loads an existing database.
 */
export function open(file: string): RrdtoolDatabase;

/**
 * Returns the current unix timestamp.
 */
export function now(): number;
```

Reply:
Sure! How do I add a default export in that case?

Reply:
I don't think you should have a default export in this case. Looking at the code as shipped at https://unpkg.com/rrdtool, there's no `export.default = ...` property, so people who don't have "esModuleInterop": true will not be able to use a default import.

Consumers of the package who want to use `import rrdtool from 'rrdtool'` will need to add `"esModuleInterop": true`, but Definitely Typed tests need to be backward compatible to show that the types will work with old code bases.

Reply:
Aah okay, thanks for the explanation!